File: processing-experiments/src/main/java/art/studio/Summer005.java
package art.studio;

import processing.core.PApplet;
import processing.core.PGraphics;

import java.util.Random;
import java.util.ArrayList;

//practice shades
public class Summer005 extends PApplet {
    int w = 1000;
    int h = 1000;
    Random alea;
    float xoff = (float) 0.0;
    float grain = (float) 0.09;
    float ang = 90;
    boolean grow = true;
    float wid = w / 10;
    ArrayList<ArrayList<Float[]>> vertices;

    @Override
    public void settings() {
        size(w, h);
    }

    @Override
    public void setup() {
        colorMode(HSB, 360, 100, 100, 250);
        alea = new Random();
        background(0, 0, 100);
        setVertices();
    }

    private void setVertices() {
        vertices = new ArrayList<>();
        for (int j = 0; j < 4; j++) {
            ArrayList<Float[]> v = new ArrayList<>();
            float x = noise(xoff) * j;
            xoff += grain;
            while (x < w) {
                Float[] one = { x, (float) 0.9 * h - j * h / 100 };
                v.add(one);
                x += noise(xoff) * w / 4;
                xoff += grain;
            }
            vertices.add(v);
        }
    }
    private void addVertices() {
        int j=vertices.size();
            ArrayList<Float[]> v = new ArrayList<>();
            float x = noise(xoff) * j;
            xoff += grain;
            while (x < w) {
                Float[] one = { x, (float) 0.9 * h - j * h / 100 };
                v.add(one);
                x += noise(xoff) * w / 4;
                xoff += grain;
            }
            vertices.add(v);
    }

    @Override
    public void draw() {
        // background(42, 0, 0);
        if (frameCount < 420) {
            noFill();
            if (alea.nextFloat() < 0.21) {
                addVertices();
            } else {
                stroke(334, 90, 90, 7);
            }
            wave();
            if (grow) {
                ang += 0.1;
                wid += 0.29;
            } else {
                ang -= 0.1;
                wid -= 0.5;
            }
            if (ang >= 180) {
                grow = false;
            }
            if (ang <= 0) {
                grow = true;
            }
        } else {
            noLoop();
            save("summer005.png");
            exit();
        }
    }

    private void wave() {
        float cx, cy, cpx1, cpy1, cpx2, cpy2;
        Float[] controls;
        for (int i = 0; i < vertices.size(); i++) {
            beginShape();
            cx = vertices.get(i).get(0)[0];
            cy = vertices.get(i).get(0)[1];
            vertex(cx, cy);
            controls = drawTang(cx, cy);
            cpx2 = controls[2];
            cpy2 = controls[3];
            for (int j = 1; j < vertices.get(i).size(); j++) {
                cx = vertices.get(i).get(j)[0];
                cy = vertices.get(i).get(j)[1];
                controls = drawTang(cx, cy);
                cpx1 = controls[0];
                cpy1 = controls[1];
                bezierVertex(cpx2, cpy2, cpx1, cpy1, cx, cy);
                cpx2 = controls[2];
                cpy2 = controls[3];
            }
            endShape();
        }
    }

    private Float[] drawTang(float cx, float cy) {
        float dx1 = cx + wid * cos(radians(ang));
        float dy1 = cy + wid * sin(radians(ang));
        float dx2 = cx + wid * cos(radians(ang + 180));
        float dy2 = cy + wid * sin(radians(ang + 180));
        Float[] res = { dx1, dy1, dx2, dy2 };
        return res;
    }

    public static void main(String[] args) {
        String[] processingArgs = { "Summer005" };
        Summer005 mySketch = new Summer005();
        PApplet.runSketch(processingArgs, mySketch);
    }

}

File: p5-experiments/plottable/summer005.js
function savesvg() {
    save("summer005.svg");
}

function savepng() {
    save("summer005.png");
}

var font
var fSize = 12
var alea = []
function preload() {
    font = loadFont("../fonts/1CAMBam_Stick_9.ttf");
    sourcecode = loadStrings('summer005.js');
}

function draw() {
    background(0, 0, 100)
    drawpostcard(globalmargin, globalmargin)
    drawpostcard(globalmargin + wpostcard, globalmargin)
    drawpostcard(globalmargin, globalmargin + hpostcard)
    drawpostcard(globalmargin + wpostcard, globalmargin + hpostcard)
    noLoop()
}

function drawpostcard(x, y) {
    noFill()
    setmargins(x, y)
    onepiece()
    rect(x, y, wpostcard, hpostcard)
    sign()
}
function onepiece() {
    alea = []
    punch4()
}

function punch() {
    stroke(0, 0, 0)
    var xoff, yoff, i, j, yinc
    yinc = 1 + random()
    xoff = floor(actualwidth * 0.01)
    i = leftmargin
    while (i < rightmargin - xoff) {
        j = bottommargin
        yoff = floor(actualheight * 0.001)
        while (j > topmargin - yoff) {
            var black = random(); alea.push(black)
            if (black < 0.21) {
                colorie(i, j, i + xoff, j, i + xoff, j + yoff, i, j + yoff)
            }
            else {
                var bound = random(); alea.push(bound)
                if (bound > 0.42) {
                    line(i + xoff, j, i, j + yoff)
                }
                if (bound < 0.1) {
                    colorie(i, j, i + xoff, j, i, j + yoff, i - xoff, j + yoff)
                }
            }
            yoff++;
            j -= yoff
        }
        xoff++
        i += xoff
    }
}

var offset = 0.0
var xinc = 0.01
function punch2() {
    stroke(0, 0, 0)
    var xoff, yoff, i, j, plein
    xoff = floor(actualwidth * 0.01)
    i = leftmargin
    while (i < rightmargin - xoff) {
        j = bottommargin
        yoff = noise(offset) * 11; offset += xinc; alea.push(yoff)
        while (j > topmargin - yoff) {
            plein = random(); alea.push(plein)
            if (plein < 0.11) {
                colorie(i, j, i + xoff, j, i + xoff, j + yoff, i, j + yoff)
            }
            else {
                var bound = random(); alea.push(bound)
                if (bound > 0.42) {
                    line(i, j, i, j + yoff)
                }
                else {
                    line(i + xoff, j, i, j + yoff)
                }
            }
            j -= yoff
            yoff = noise(offset) * 11; offset += xinc; alea.push(yoff)
        }
        xoff++
        i += xoff
    }
}

function punch3() {
    var ox1, oy1, dx1, dy1, ox2, oy2, dx2, dy2, t1
    for (t1 = 0; t1 < 1; t1 += 0.05) {
        ox1 = leftmargin
        oy1 = bottommargin
        dx1 = rightmargin
        dy1 = (1 - t1) * bottommargin + (t1 * topmargin)
        ox2 = leftmargin
        oy2 = bottommargin
        dx2 = rightmargin
        dy2 = (1 - (t1+0.05)) * bottommargin + ((t1+0.05) * topmargin)
        section(ox1,oy1,dx1,dy1,dx2,dy2)
    }
    for (t1 = 0; t1 < 1; t1 += 0.05) {
        ox1 = leftmargin
        oy1 = bottommargin
        dx1 = (1 - t1) * leftmargin + (t1 * rightmargin)
        dy1 = topmargin
        ox2 = leftmargin
        oy2 = bottommargin
        dx2 = (1 - (t1+0.05)) * leftmargin + ((t1+0.05) * rightmargin)
        dy2 = topmargin
        section(ox1,oy1,dx1,dy1,dx2,dy2)
    }
    ellipse(rightmargin,topmargin,7,7)
}

function section(x1, y1, x2, y2, x3, y3){
    var ox1, oy1, dx1, dy1, ox2, oy2, dx2, dy2, t1, step
    step=0.01
    for (t1 = 0; t1 < 1; t1 += step) {
        ox1 = (1 - t1) * x1 + (t1 * x2)
        oy1 = (1 - t1) * y1 + (t1 * y2)
        dx1 = (1 - t1) * x1 + (t1 * x3)
        dy1 = (1 - t1) * y1 + (t1 * y3)
        if(random()<0.42){line(ox1,oy1,dx1,dy1)}
        ox2 = (1 - (t1-step)) * x1 + ((t1-step) * x2)
        oy2 = (1 - (t1-step)) * y1 + ((t1-step) * y2)
        dx2 = (1 - (t1-step)) * x1 + ((t1-step) * x3)
        dy2 = (1 - (t1-step)) * y1 + ((t1-step) * y3)
        if(random()<0.2){colorie(dx2,dy2,dx1,dy1,ox1,oy1,ox2,oy2)}
    }
    for (t1 = 0; t1 < 1; t1 += step) {
        ox1 = (1 - t1) * x1 + (t1 * x2)
        oy1 = (1 - t1) * y1 + (t1 * y2)
        dx1 = (1 - (t1+step)) * x1 + ((t1+step) * x2)
        dy1 = (1 - (t1+step)) * y1 + ((t1+step) * y2)
        if(random()>0.42){line(ox1,oy1,dx1,dy1)}
    }
}


var noff=42
var ninc=0.001
function punch4(){
    var x1=leftmargin
    var x2=leftmargin
    var y2=topmargin
    var step=1
    while(x1<rightmargin){
        ocean(x2,bottommargin,x1,y2)
        x1+=random(2,9)
        x2+=2
        y2+=2
        step+=random()
    }
    var cx=leftmargin+actualwidth*0.5
    var cy=topmargin+actualheight*0.5

}

function ocean(x1, y1, x2, y2){
    var ox1, oy1, cpx1,cpy1,cpx2,cpy2,t1, step
    var ratio1=0.5
    var ratio2=0.5
    step=0.2
    beginShape()
    ox1=x1
    oy1=y1
    vertex(x1,y1)
    for (var t1 = step; t1 < 1+step; t1 += step) {
        dx1 = (1 - t1) * x1 + (t1 * x2)
        dy1 = (1 - t1) * y1 + (t1 * y2)
        cpx1 = ((1 - ratio1) * ox1 + (ratio1 * dx1))-21//+random(-42,84)
        cpy1 = ((1 - ratio1) * oy1 + (ratio1 * dy1))
        cpx2 = ((1 - ratio2) * ox1 + (ratio2 * dx1))+21//+random(-42,84)
        cpy2 = ((1 - ratio2) * oy1 + (ratio2 * dy1))
        bezierVertex(cpx1,cpy1,cpx2,cpy2,dx1, dy1)
        ox1 = dx1
        oy1 = dy1
    }
    endShape()
}

function colorie(x1, y1, x2, y2, x3, y3, x4, y4) {
    var ox, oy, dx, dy
    push()
    var red = random(); alea.push(red)
    if (red < 0.1) { stroke(0, 100, 100) }
    for (var t1 = 0; t1 < 1; t1 += 0.03) {
        ox = (1 - t1) * x1 + (t1 * x2)
        oy = (1 - t1) * y1 + (t1 * y2)
        dx = (1 - t1) * x4 + (t1 * x3)
        dy = (1 - t1) * y4 + (t1 * y3)
        line(ox, oy, dx, dy)
    }
    pop()
}


function sign() {
    var title = "[summer005]"
    var credits = "al.my.re :: July 2024 (p5.js ~ CamBam Stick ~ gcode ~ juicy ~ " + alea.length + " random numbers)"
    textFont(font)
    textSize(fSize)
    push()
    translate(rightmargin, bottommargin);
    rotate(radians(270))
    showcredits(0, fSize, title, credits)
    pop()
}
